Content.Shared/Whitelist/EntityWhitelistSystem.Prototype.cs

```diff
@@ -0,0 +1,139 @@
+using Content.Shared.Item;
+using Content.Shared.Tag;
+using Robust.Shared.Prototypes;
+
+namespace Content.Shared.Whitelist;
+
+public sealed partial class EntityWhitelistSystem
+{
+    /// <summary>
+    /// Checks whether a given entity prototype satisfies a whitelist.
+    /// </summary>
+    public bool IsValid(EntityWhitelist list, [ForbidLiteral] EntProtoId protoId)
+    {
+        var prototype = _proto.Index(protoId);
+
+        if (list.Components != null)
+        {
+            foreach (var comp in list.Components)
+            {
+                if (prototype.Components.ContainsKey(comp))
+                {
+                    if (!list.RequireAll)
+                        return true;
+                }
+                else if (list.RequireAll)
+                    return false;
+            }
+        }
+
+        if (list.Sizes != null && prototype.TryGetComponent<ItemComponent>(out var itemComp, Factory))
```

Comment:
Minor performance thing, but you should probably just cache `factory.GetComponentName<T>(ItemComponent)` and then re-use that with `TryGetComponent(string, name, out T Comp)` instead of using the overload that takes in the factory. Same applies to the tag component below.